app/plugin_api.go

```diff
@@ -914,3 +914,59 @@ func (api *PluginAPI) PluginHTTP(request *http.Request) *http.Response {
 	api.app.ServeInterPluginRequest(responseTransfer, request, api.id, destinationPluginId)
 	return responseTransfer.GenerateResponse()
 }
+
+func (api *PluginAPI) CreateSlashCommand(cmd *model.Command) (*model.Command, *model.AppError) {
+	cmd.CreatorId = ""
+	cmd.PluginId = api.id
+
+	cmd, appErr := api.app.CreateCommand(cmd)
+
+	return cmd, appErr
+}
+
+func (api *PluginAPI) ListSlashCommands(teamID string, customOnly bool) ([]*model.Command, *model.AppError) {
```

Comment:
For the customOnly flag, Christopher would usually call it with `customOnly` as `true`, but he wants to keep it as general as possible. You could argue that returning everything keeps it as general as possible, but then you'd have to add even more code in the plugin-api to re-narrow down the results. To me it makes more sense to narrow it with the customOnly flag, which the api already supports. Then we could simplify it in the plugin-api, as you suggest.

About returning custom commands from every team -- we could add that in the plugin-api, imo. Considering how many teams some orgs have, it would be a non-trivial time cost for that call, unless you wrote a new SQL query. Did you want to do that now? (I would vote for leaving that for later, if/when we need it, and we find that calling this per team was costing too much.)

Reply:
@crspeller, I'm still hesitant to include `customOnly` as a filter parameter. As it stands right now, I feel like this API is trying to be too much like its REST equivalent, which is clearly geared towards the need for autocomplete. If we were standing up this API from scratch, would we support a `customOnly`? Why can't I query `pluginOnly` or `builtInOnly`? My inclination is still just to simplify this to return all, and support whatever filtering we need in the plugin-api wrapper instead.

I'm fine with keeping the `teamId`, since that's a fundamental aspect of how commands work and this is probably not the right layer to change that. We can definitely kick that out from our plans altogether right now.

Reply:
Quick question -- is there any way to filter (from a list of commands) plugin commands vs built-in commands? Custom commands have URLs, so that's possible.

If there isn't, then I think we should add essentially a bitmask to the list commands:
```
type ListCommandsOptions struct {
  builtIn bool
  plugin bool
  custom bool
}
```

Reply:
That's a good question -- do plugin commands have an `Id`, given that we don't actually persist these to the database?

Reply:
@lieut-data looks like there's no difference between built-in and plugin commands, they both look the same when they're in a list of commands. I think we're going to have to use something like an options. Any other ideas, or should I just do that?

Reply:
I'm fine with keeping things simple -- I guess I'm actually looking to make things /simpler/ re: even supporting `customOnly`. Given this API comes as an ask from @crspeller, I'm fine deferring to his requirements.

Reply:
The new ListXCommands api fixes this in a way that makes a bit more sense, I think.